src/rust/engine/fs/store/src/lib.rs

```diff
@@ -1297,17 +1343,56 @@ impl Store {
                   Permissions::Writable if f.is_executable() => 0o755,
                   Permissions::Writable => 0o644,
                 };
-                store.materialize_file(path, f.digest(), mode).await
+
+                if can_be_immutable && f.digest().size_bytes > IMMUTABLE_FILE_SIZE_LIMIT {
+                  let dest_path = immutable_inputs
+                    .unwrap()
+                    .path_for_file(f.digest(), mode)
+                    .await?;
+                  store
+                    .materialize_symlink(path, dest_path.to_str().unwrap().to_string())
+                    .await
+                } else {
+                  store.materialize_file(path, f.digest(), mode).await
+                }
               }
               directory::Entry::Symlink(s) => {
                 store
                   .materialize_symlink(path, s.target().to_str().unwrap().to_string())
                   .await
               }
-              directory::Entry::Directory(_) => {
-                store
-                  .materialize_directory_helper(path, false, parent_to_child, perms)
-                  .await
+              directory::Entry::Directory(d) => {
+                // NB: This overcounts the size of the digest, since it includes the size of directory digests.
+                if can_be_immutable
+                  && d
+                    .tree()
+                    .digests()
+                    .iter()
+                    .map(|d| d.size_bytes)
+                    .sum::<usize>()
+                    > IMMUTABLE_DIR_SIZE_LIMIT
```

Comment:
This will end up re-computing this value at each directory node that is materialized, which will revisit digests a bunch of times... `O(n^2)` I think?

It would be better to compute the directory sizes once up front using a memoized walk... maybe via a new method of `DigestTrie` that did a recursive memoized walk to build a `HashMap<Digest, usize>`.